src/clean-core/string.hh

```diff
@@ -0,0 +1,317 @@
+
+    char const* c_str() const { return data(); }
+
+    char* begin() { return data(); }
+    char const* begin() const { return data(); }
+    char* end() { return data() + size(); }
+    char const* end() const { return data() + size(); }
+
+    size_t size() const { return _is_short() ? _rep.s.size >> 1 : _rep.l.size; }
+    size_t capacity() const { return _is_short() ? sbo_capacity - 1 : _rep.l.capacity; }
+
+    bool empty() { return size() == 0; }
+
+    char& front()
+    {
+        CC_CONTRACT(size() > 0);
+        return data()[0];
+    }
+    char const& front() const
+    {
+        CC_CONTRACT(size() > 0);
+        return data()[0];
+    }
+
+    char& back()
+    {
+        CC_CONTRACT(size() > 0);
+        return data()[size() - 1];
+    }
+    char const& back() const
+    {
+        CC_CONTRACT(size() > 0);
+        return data()[size() - 1];
+    }
+
+    char& operator[](size_t i)
+    {
+        CC_CONTRACT(i < size());
+        return data()[i];
+    }
+    char const& operator[](size_t i) const
+    {
+        CC_CONTRACT(i < size());
+        return data()[i];
+    }
+
+    // ctors
+public:
+    string() { _rep.r = {}; }
+
+    // TODO: more constructors (char const*, string_view)
+
+    [[nodiscard]] static string uninitialized(size_t size)
+    {
+        string s;
+        if (size < sbo_capacity)
+            s._rep.s.size = short_flag + (size << 1);
+        else
+        {
+            s._rep.l.size = size;
+            s._rep.l.capacity = size;
+            s._rep.l.data = new char[size];
+        }
+        return s;
+    }
+
+    [[nodiscard]] static string filled(size_t size, char value)
+    {
+        string s;
+        s.resize(size, value);
+        return s;
+    }
+
+    string(string const& rhs)
+    {
+        if (rhs._is_short())
+            _rep.r = rhs._rep.r;
+        else if (rhs._rep.l.data == nullptr)
+            _rep.r = {};
+        else
+        {
+            _rep.l.data = new char[rhs._rep.l.size];
+            std::memcpy(_rep.l.data, rhs._rep.l.data, rhs._rep.l.size);
+            _rep.l.size = rhs._rep.l.size;
+            _rep.l.capacity = rhs._rep.l.size;
+        }
+    }
+    string(string&& rhs) noexcept
+    {
+        _rep.r = rhs._rep.r;
+        rhs._rep.r = {};
+    }
+
+    string& operator=(string const& rhs)
+    {
+        if (this != &rhs)
+        {
+            if (!_is_short())
+                delete[] _rep.l.data;
+
+            if (rhs._is_short())
+                _rep.r = rhs._rep.r;
+            else if (rhs._rep.l.data == nullptr)
+                _rep.r = {};
+            else
+            {
+                _rep.l.data = new char[rhs._rep.l.size];
+                std::memcpy(_rep.l.data, rhs._rep.l.data, rhs._rep.l.size);
+                _rep.l.size = rhs._rep.l.size;
+                _rep.l.capacity = rhs._rep.l.size;
+            }
+        }
+        return *this;
+    }
+    string& operator==(string&& rhs) noexcept
```

Comment:
Should this be an assignment operator instead?

Reply:
:D nice